### DISCUSSION TITLE
How should we read values from an `Effect` in the future.

### DISCUSSION THREAD

**Discussion by notcome** [pointfreeco/swift-composable-architecture#1925] [Q&A] [answered]
So I decided to live on the edge and try to migrate my codebase to `prerelease/1.0`. One thing I noticed that the conformance to `Publisher` is now deprecated. However at the moment, we don't have any way to read values from an `Effect`, other than treating it as a Combine `Publisher`. What should we do instead? Will `Effect` become an `AsyncSequence`?

**Comment by tgrapperon** [comment, marked as answer]
Hey @notcome! Conflating `Effect` with `Publisher<_, Never>` is a byproduct of `Effect` being based on `Publisher`. Ideally, your dependencies should produce some `Publisher<Output, Failure>`. In other words, the dependency doesn't produce an `Effect`, it produces a `Publisher` of events. Your reducer can then use this publisher to transform this sequence of events into a sequence of `Action`s (that is, an `Effect`). You can't really read values from an `Effect` because an `Effect` isn't strictly speaking something that produces values[^1]. This is something that is expressed by the `Store`.

Once you keep this dichotomy in mind, the conversion is simple: The publisher of events becomes an async sequence of events, and your reducer can iterate over this sequence of events to produce actions for the store. There is an [example of such conversion in `EffectPublisher`'s documentation](https://github.com/pointfreeco/swift-composable-architecture/blob/f7c75217a8087167aacbdad3fe4950867f468a52/Sources/ComposableArchitecture/Effects/Publisher.swift#L63-L79).

To sum up, pre-`1.0`, `Effect` is a `Publisher` and a `Publisher` of `Action`s that can't fail is an `Effect`, but in `1.0`, `Effect` is an opaque type and a `Publisher` of `Action`s that can't fail is not an `Effect` (but it can used to create one `Effect`). Pre-`1.0`, A `Publisher<Output, Never>` is technically an `Effect<Output>`, but this doesn't make a lot of sense if `Output` is not the `Action` for some `Store`, as `Effect`'s only purpose is to be consumed by a `Store`.

If you encounter issues while modernizing your codebase, feel free to post a few problematic bits, so we can provide more specific orientations!

[^1]: It is right now because `Effect`s are `Publisher`, but this will not be the case anymore in `1.0`.

**Reply by notcome** [reply]
Thanks! I think that made lots of sense. Lately my code was abusing `Effect` as a replacement for `AnyPublisher`, and that probably isn't a very good idea. I used it to read update from Apple and non-Apple APIs. I should probably move it to `AsyncSequence`, using things like `AsyncChannel` instead.

**Reply by tgrapperon** [reply]
I'm not certain that you need `AsyncChannel`, but yes, modelling your dependencies as `AsyncSequence`s is probably a good move. You can easily convert any publisher to an async sequence using the `.values` property starting from `iOS 15`, as shown in `EffectPublisher`'s documentation.
Please note that the de-facto erasure for `AsyncSequence` in TCA is to wrap it in an `AsyncStream` (`Dependencies` provides an `eraseToStream()` method that you can call on any `AsyncSequence` for this purpose).

**Reply by notcome** [reply]
> I'm not certain that you need AsyncChannel

I was abusing `Effect.run` to provide implementations for my dependencies. So converting them to `AsyncChannel` feels natural, based on my limited experience with `swift-async-algorithms`, 😂. But certainly in many cases I can do it the functional style.
